### PR TITLE
Feature: JWT parser

### PR DESCRIPTION
**Pull request by doncem**
### JIRA link (if applicable) ###

[Private Beta Users](https://tools.hmcts.net/jira/browse/RPE-428)

### Change description ###

Introducing JWT token parser service

- [x] filter implementation with placeholder for jwt token parser
- [x] implement jwt token parser
- [ ] stick filter in security config and write bunch of tests

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/feature/service/JwtParserService.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    public JwtParserService(JwtParser parser) {
+        this.parser = parser;
+    }
+
+    public UserTokenDetails parse(String jwtToken) {
+        try {
+            Claims claims = parser.parseClaimsJws(jwtToken).getBody();
+
+            String id = claims.get("id").toString();
+            List<String> roles = Arrays.asList(claims.get("data").toString().split(","));
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/feature/service/JwtParserService.java]: should this claim not be called `roles`?

File: src/main/java/uk/gov/hmcts/reform/feature/service/JwtParserService.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+            return new UserTokenDetails(id, roles);
+        } catch (Exception exception) {
+            // voiding any exception
+            return null;
```
Comment (rkondratowicz) [inline src/main/java/uk/gov/hmcts/reform/feature/service/JwtParserService.java]: add logging.
Also, shall we continue at this stage, or throw a custom exception?

File: src/test/java/uk/gov/hmcts/reform/feature/service/JwtParserServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    private static final String SIGNING_KEY = "signing";
+
+    private final JwtParserService service = new JwtParserService(Jwts.parser().setSigningKey(SIGNING_KEY));
+
+    @Test
+    public void should_parse_valid_jwt_token_and_get_user_details() {
+        // given
+        String jwtToken = Jwts.builder()
+            .signWith(SignatureAlgorithm.HS256, SIGNING_KEY)
+            .addClaims(ImmutableMap.of("id", "1234", "data", "beta,test"))
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/feature/service/JwtParserServiceTest.java]: as before, idam seems to be putting roles in `roles` claim
https://github.com/hmcts/idam-app/blob/master/api/src/main/java/uk/gov/hmcts/idam/infrastructure/security/authentication/UserAuthenticator.java#L287-L301

File: src/test/java/uk/gov/hmcts/reform/feature/service/JwtParserServiceTest.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+        assertThat(service.parse(jwtToken)).isNull();
+    }
+
+    @Test
+    public void should_fail_to_parse_jwt_token_when_wrong_jwt_string_passed_and_return_null() {
+        assertThat(service.parse("jtwtoken.with2dots.init")).isNull();
+    }
+
+    @Test
+    public void should_fail_to_parse_jwt_token_when_wrong_string_passed_and_return_null() {
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/feature/service/JwtParserServiceTest.java]: fail or return null?

File: src/test/java/uk/gov/hmcts/reform/feature/service/JwtParserServiceTest.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+import org.junit.Test;
+import uk.gov.hmcts.reform.feature.model.UserTokenDetails;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class JwtParserServiceTest {
+
+    private static final String SIGNING_KEY = "signing";
+
+    private final JwtParserService service = new JwtParserService(Jwts.parser().setSigningKey(SIGNING_KEY));
```
Comment (rkondratowicz) [inline src/test/java/uk/gov/hmcts/reform/feature/service/JwtParserServiceTest.java]: is the signing key required?

We will not have access to IDAM's signing key

File: src/main/java/uk/gov/hmcts/reform/feature/service/JwtParserService.java (hunk 1 above)
Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/feature/service/JwtParserService.java]: no it should not: https://github.com/hmcts/idam-app/blob/master/api/src/main/java/uk/gov/hmcts/idam/infrastructure/security/authentication/ClaimProperties.java#L4

File: src/main/java/uk/gov/hmcts/reform/feature/service/JwtParserService.java (hunk 2 above)
Comment (doncem) [reply src/main/java/uk/gov/hmcts/reform/feature/service/JwtParserService.java]: will add.

yes, we should continue no matter what. it is not application nor any other error. always assume as nothing wrong happened. moreover - one should always get a response about feature and filter just provides a way of retrieving user roles

File: src/test/java/uk/gov/hmcts/reform/feature/service/JwtParserServiceTest.java (hunk 4 above)
Comment (doncem) [reply src/test/java/uk/gov/hmcts/reform/feature/service/JwtParserServiceTest.java]: `fail to parse` and then `return null`

**Comment by doncem** [comment]
Due to lack of proper user management among the services more direct approach is needed. Canning this even though implementation is pretty much correct

**Pull request closed without merging**